### DISCUSSION TITLE
Question - How use flatsharp with Unity

### DISCUSSION THREAD

**Discussion by MateusMendesSantana** [jamescourtney/FlatSharp#294] [General]
Hello, can you help me? How can i generate Schema for unity, can i use the same schema generated in .Net Core or generate for Unity? Is there a .unitypackage for Unity? It took me a while to know how to access the .serializer of the schemas, they are only generated when the "(fs_serializer)" is added to the tables, may.

`table UpdatePlaceSchema (fs_serializer) {
    name: string;
    description: string;
    maxAvatars: int;
}`

**Comment by jamescourtney** [comment]
Hello! Thanks for reaching out.

The same schema should work for both .NET Core and Unity. However, to use the schema in Unity, you may need to take some special steps. I've seen people use FlatSharp with unity successfully with the following patterns:

- A prebuild step where you invoke the FlatSharp compiler manually using the command line tool (included as part of the `FlatSharp.Compiler` nuget package). Command line usage can be found here: https://github.com/jamescourtney/FlatSharp/wiki/Compiler
- A .NET standard project (ie, MSBuild) which uses FlatSharp normally. Then you can reference that generated assembly from Unity.

There is no `.unitypackage` for FlatSharp. I never expected FlatSharp to be used by so many folks in the Unity community, and I just don't have any expertise. If you would be willing to help, I'd love to publish an official Unity FlatSharp package.

**Comment by MateusMendesSantana** [comment]
Tks, currently the unity package manager supports installing plugins from a repository on github, maybe this would be a better alternative instead of generating a .unitypackage . I tried to use this feature, but for some reason it didn't work.
![image](https://user-images.githubusercontent.com/35671729/167909202-def1ab4c-1dfe-48d2-9244-5c884c1426c9.png)
Has anyone posted any unity pre build code? If not, I'll produce one and post it, so that people can use it as a base in the future.

**Comment by jamescourtney** [comment]
No, I don't think anyone has done that. Anything you could contribute to streamlining Unity onboarding would be greatly appreciated!

**Reply by vvuk** [reply]
I may be able to take this on -- it should be possible to package everything (including `flatc` and the Compiler) as a package, and have it be automatically invoked when fbs files change to generate C#.

**Reply by jamescourtney** [reply]
That would be awesome! And any way you could help me figure out how to get a simple unity environment set up would be really helpful. I'd love to have explicit support for unity scenarios.

**Comment by MateusMendesSantana** [comment]
Hey, after some tests I got some information, which can help other people. Process to unity integration:
- Clone/Download Github FlatSharp project or download release.
- Open project on Visual Studio.
- Edit .csproject from Compiler and Runtime.
- Go to and .csproject edit and add **netstandard2.1** to target plataform.
 `<TargetFrameworks>netstandard2.0;netstandard2.1;netcoreapp3.1;net5.0;net6.0;**netstandard2.1**</TargetFrameworks>`
- Copy both generated DLLs to Unity Asset folder.
- Generate schemas from .fbs inside Unity project.
`dotnet C:\...\.nuget\packages\flatsharp.compiler\6.3.1\tools\net6.0\FlatSharp.Compiler.dll -i C:\...\schemas.fbs -o C:\...t\Assets\Scripts`

I do not recommend a pre build script, because if the code is broken it will not be possible to update the schemas.
